File: workspace/solutions/java/lib/src/test/java/utils/ExampleSet.java
// Copyright (C) 2024 Zuoqiu Yingyi
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU Affero General Public License as
// published by the Free Software Foundation, either version 3 of the
// License, or (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU Affero General Public License for more details.
//
// You should have received a copy of the GNU Affero General Public License
// along with this program.  If not, see <https://www.gnu.org/licenses/>.

package utils;

import static org.junit.jupiter.api.Assertions.assertEquals;

import java.nio.file.Files;
import java.nio.file.Path;
import java.nio.file.Paths;
import java.util.ArrayList;
import java.util.List;

import com.alibaba.fastjson2.JSON;
import com.alibaba.fastjson2.JSONArray;
import com.alibaba.fastjson2.JSONObject;

/**
 * 示例集合 (相同初始化参数的一组示例)
 */
public class ExampleSet {
    // System.getProperty("user.dir") -> workspace/solutions/java/lib
    public static final String SOLUTIONS_TEST_EXAMPLES_DIRECTORY = "./../../../packages/examples";

    /**
     * 将包名转换为题解 ID
     * @param packageName 包名 (格式: "_0._1._2._3")
     * @return 题解 ID (格式: 0123)
     */
    public static String package2id(final String packageName) {
        return packageName
            .replace("s", "")
            .replace(".", "");
    }

    /**
     * 将包名转换为路径
     * @param packageName 包名 (格式: "_0._1._2._3")
     * @return 路径 (格式: "0/1/2/3")
     */
    public static final String package2path(final String packageName) {
        return packageName
            .replace("s", "")
            .replace(".", "/");
    }

    /**
     * 读取示例文件
     * @param packageName 包名 (格式: "_0._1._2._3")
     * @return JSON 字符串
     */
    public static final String readExamplesFile(final String packageName) {
        try {
            final String id = ExampleSet.package2id(packageName);
            final String path = ExampleSet.package2path(packageName);
            final Path file_path = Paths.get(ExampleSet.SOLUTIONS_TEST_EXAMPLES_DIRECTORY, path, id + ".json");
            final String json = Files.readString(file_path);
            return json;
        }
        catch (Exception e) {
            return null;
        }
    }

    /**
     * 创建示例列表
     * @param json 示例 JSON 字符串
     * @param constructorParameterTypeNames 构造函数参数类型名称列表
     * @param methodParameterTypeNames 方法参数类型名称列表
     * @param methodReturnTypeName 方法返回值类型名称
     * @return 示例列表
     */
    public static final List<ExampleSet> fromJson(
        final String json,
        final String[] constructorParameterTypeNames,
        final String[] methodParameterTypeNames,
        final String methodReturnTypeName
    ) {
        final JSONArray examples_json_array = JSON.parseArray(json);
        final List<ExampleSet> exampleSetList = new ArrayList<ExampleSet>();
        if (constructorParameterTypeNames.length > 0) { // 构造函数有参数
            final List<ExampleWithInit> examplesWithInit = JSON.parseArray(json, ExampleWithInit.class);
            final int examples_count = examplesWithInit.size();
            for (int i = 0; i < examples_count; ++i) {
                final JSONObject example_json = examples_json_array.getJSONObject(i);
                final ExampleWithInit example = examplesWithInit.get(i);

                /* 根据类型解析 JSON 对象 */
                example.setInit(example_json, constructorParameterTypeNames);
                example.setInputs(example_json, methodParameterTypeNames);
                example.setOutputs(example_json, methodReturnTypeName);

                assertEquals(
                    example.init.length,
                    constructorParameterTypeNames.length,
                    "example.init.length != constructor parameters length"
                );
                assertEquals(
                    example.inputs.length,
                    example.outputs.length,
                    "example.inputs.length != example.outputs.length"
                );

                /* 根据 inputs 与 outputs 构造 examples */
                final int examples_length = example.inputs.length;
                final List<Example> examples = new ArrayList<Example>(examples_length);
                for (int j = 0; j < examples_length; j++) {
                    examples.addLast(new Example(example.inputs[j], example.outputs[j]));
                }

                exampleSetList.add(new ExampleSet(
                    example.init,
                    examples
                ));
            }
        }
        else { // 构造函数无参数
            final List<Example> examples = JSON.parseArray(json, Example.class);
            final int examples_count = examples.size();
            for (int i = 0; i < examples_count; ++i) {
                final JSONObject example_json = examples_json_array.getJSONObject(i);
                final Example example = examples.get(i);

                /* 根据类型解析 JSON 对象 */
                example.setInput(example_json, methodParameterTypeNames);
                example.setOutput(example_json, methodReturnTypeName);
            }
            exampleSetList.addLast(new ExampleSet(new Object[0], examples));
        }
        return exampleSetList;
    }

    /**
     * 解析 JSON 列表
     * @param jsonArray JSON 列表
     * @param typeNames 类型名称列表
     */
    public static Object[] jsonList(
        final JSONArray jsonArray,
        final String[] typeNames
    ) {
        final int array_length = jsonArray.size();
        assertEquals(
            array_length,
            typeNames.length,
            "Number of JSON Array != Number of Types"
        );

        final Object[] array = new Object[jsonArray.size()];
        for (int i = 0; i < array_length; ++i) {
            switch (typeNames[i]) {
                case "String":
                    array[i] = jsonArray.getString(i);
                    break;
                case "int":
                    array[i] = jsonArray.getIntValue(i);
                    break;
                case "int[]":
                    array[i] =
                        jsonArray
                            .getJSONArray(i)
                            .toJavaList(Integer.class)
                            .stream()
                            .mapToInt(Integer::intValue)
                            .toArray();
                    break;

                default:
                    array[i] = jsonArray.get(i);
                    break;
            }
        }
        return array;
    }

    /**
     * 解析 JSON 值
     * @param jsonObject JSON 对象
     * @param valueKey 字段名称
     * @param typeName 类型名称
     */
    public static Object jsonValue(
        final JSONObject jsonObject,
        final String valueKey,
        final String typeName
    ) {
        switch (typeName) {
            case "String":
                return jsonObject.getString(valueKey);
            case "int":
                return jsonObject.getIntValue(valueKey);
            case "int[]":
                return jsonObject
                    .getJSONArray(valueKey)
                    .toJavaList(Integer.class)
                    .stream()
                    .mapToInt(Integer::intValue)
                    .toArray();
            default:
                return jsonObject.get(valueKey);
        }
    }

    /**
     * 题解构造函数参数
     */
    public Object[] init;

    /**
     * 题解输入输出示例列表
     */
    public List<Example> examples;

    public ExampleSet(
        final Object[] init,
        final List<Example> examples
    ) {
        this.init = init;
        this.examples = examples;
    }
}

File: workspace/solutions/kotlin/lib/src/test/kotlin/utils/ExampleSet.kt
// Copyright (C) 2024 Zuoqiu Yingyi
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU Affero General Public License as
// published by the Free Software Foundation, either version 3 of the
// License, or (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU Affero General Public License for more details.
//
// You should have received a copy of the GNU Affero General Public License
// along with this program.  If not, see <https://www.gnu.org/licenses/>.

@file:Suppress("ktlint:standard:no-wildcard-imports")

package utils

import kotlinx.serialization.json.Json
import kotlinx.serialization.json.JsonArray
import kotlinx.serialization.json.JsonElement
import kotlinx.serialization.json.boolean
import kotlinx.serialization.json.double
import kotlinx.serialization.json.encodeToJsonElement
import kotlinx.serialization.json.int
import kotlinx.serialization.json.jsonArray
import kotlinx.serialization.json.jsonObject
import kotlinx.serialization.json.jsonPrimitive
import kotlinx.serialization.json.long
import java.io.File
import kotlin.io.path.Path
import kotlin.test.*

public class ExampleSet(
    public val init: Array<Any>,
    public val examples: List<Example>,
) {

    companion object {
        public val SOLUTIONS_TEST_EXAMPLES_DIRECTORY: String = "./../../../packages/examples"

        /**
         * 将包名转换为题解 ID
         * @param packageName 包名 (格式: "_0._1._2._3")
         * @return 题解 ID (格式: 0123)
         */
        public final fun package2id(packageName: String): String = packageName
            .replace("s", "")
            .replace(".", "")

        /**
         * 将包名转换为路径
         * @param packageName 包名 (格式: "_0._1._2._3")
         * @return 路径 (格式: "0/1/2/3")
         */
        public final fun package2path(packageName: String): String = packageName
            .replace("s", "")
            .replace(".", "/")

        /**
         * 读取示例文件
         * @param packageName 包名 (格式: "_0._1._2._3")
         * @return JSON 字符串
         */
        public final fun readExamplesFile(packageName: String): String? {
            try {
                val id = ExampleSet.package2id(packageName)
                val path = ExampleSet.package2path(packageName)
                val file_path = Path(
                    ExampleSet.SOLUTIONS_TEST_EXAMPLES_DIRECTORY,
                    path,
                    "$id.json",
                )
                val json = File(file_path.toString()).readText()
                return json
            } catch (e: Exception) {
                return null
            }
        }

        /**
         * 创建示例列表
         * @param json 示例 JSON 字符串
         * @param constructorParameterTypeNames 构造参数类型名称列表
         * @param methodParameterTypeNames 方法参数类型名称列表
         * @param methodReturnTypeName 方法返回值类型名称
         * @return 示例列表
         */
        public final fun fromJson(
            json: String,
            constructorParameterTypeNames: List<String>,
            methodParameterTypeNames: List<String>,
            methodReturnTypeName: String,
        ): List<ExampleSet> {
            val examples_json_array = Json.parseToJsonElement(json).jsonArray
            val exampleSetList: MutableList<ExampleSet> = mutableListOf()

            if (constructorParameterTypeNames.isNotEmpty()) {
                examples_json_array.forEach { example_json ->
                    val example_json_object = example_json.jsonObject
                    val init = ExampleSet.getInit(example_json_object.get("init")?.jsonArray, constructorParameterTypeNames)
                    val inputs = example_json_object.get("inputs")?.jsonArray
                    val outputs = example_json_object.get("outputs")?.jsonArray

                    assertNotNull(inputs, "example.inputs is Null")
                    assertNotNull(outputs, "example.outputs is Null")
                    assertEquals(
                        inputs.size,
                        outputs.size,
                        "example.inputs.size != example.outputs.size",
                    )

                    val example_size = inputs.size
                    val examples = (0..<example_size).map { i ->
                        Example().apply {
                            this.setInput(inputs[i].jsonArray, methodParameterTypeNames)
                            this.setOutput(outputs[i], methodReturnTypeName)
                        }
                    }
                    exampleSetList.addLast(ExampleSet(init, examples))
                }
            } else { // 构造函数无参数
                val examples = examples_json_array.map { example_json_element ->
                    val example_json_object = example_json_element.jsonObject
                    Example().apply {
                        this.setInput(example_json_object.get("input")?.jsonArray, methodParameterTypeNames)
                        this.setOutput(example_json_object.get("output"), methodReturnTypeName)
                    }
                }
                exampleSetList.addLast(ExampleSet(arrayOf(), examples))
            }
            return exampleSetList
        }

        /**
         * 将 JSON 元素转换为指定类型
         * @param jsonElement JSON 元素
         * @param typeName 类型名称
         * @return 指定类型
         *
         * @see [kotlinx.serialization.json](https://kotlinlang.org/api/kotlinx.serialization/kotlinx-serialization-json/kotlinx.serialization.json/)
         */
        public final fun json2any(
            jsonElement: JsonElement,
            typeName: String,
        ): Any = when (typeName) {
            "kotlin.String" -> jsonElement.jsonPrimitive.content
            "kotlin.Boolean" -> jsonElement.jsonPrimitive.boolean
            "kotlin.Int" -> jsonElement.jsonPrimitive.int
            "kotlin.Long" -> jsonElement.jsonPrimitive.long
            "kotlin.Double" -> jsonElement.jsonPrimitive.double
            "kotlin.IntArray" -> jsonElement.jsonArray.map { it.jsonPrimitive.int }.toIntArray()
            "kotlin.Array<kotlin.IntArray>" -> jsonElement.jsonArray.map { it.jsonArray.map { it.jsonPrimitive.int }.toIntArray() }.toTypedArray()
            else -> throw IllegalArgumentException("Unsupported type (json2any): $typeName")
        }

        /**
         * 将 Any 类型的值转换为指定类型
         * @param value 任意类型的值
         * @param typeName 类型名称
         * @return 指定类型
         */
        public final fun any2json(
            value: Any,
            typeName: String,
        ): JsonElement = when (typeName) {
            "kotlin.String" -> Json.encodeToJsonElement(value as String)
            "kotlin.Boolean" -> Json.encodeToJsonElement(value as Boolean)
            "kotlin.Int" -> Json.encodeToJsonElement(value as Int)
            "kotlin.Long" -> Json.encodeToJsonElement(value as Long)
            "kotlin.Double" -> Json.encodeToJsonElement(value as Double)
            "kotlin.IntArray" -> Json.encodeToJsonElement(value as IntArray)
            "kotlin.Array<kotlin.IntArray>" -> Json.encodeToJsonElement(value as Array<IntArray>)
            else -> throw IllegalArgumentException("Unsupported type (any2json): $typeName")
        }

        /**
         * 将 JSON 元素转换为 Array 类型
         * @param jsonArray JSON 列表元素
         * @param typeNames 类型名称列表
         * @return 指定类型
         */
        public final fun json2array(
            jsonArray: JsonArray,
            typeNames: List<String>,
        ): Array<Any> = List(jsonArray.size) { index ->
            ExampleSet.json2any(jsonArray.get(index), typeNames[index])
        }.toTypedArray()

        /**
         * 设置题解示例输入
         * @param initJsonArray 构造函数参数值列表
         * @param parameterTypeNames 构造函数参数类型列表
         */
        public final fun getInit(
            initJsonArray: JsonArray?,
            parameterTypeNames: List<String>,
        ): Array<Any> {
            assertNotNull(
                initJsonArray,
                "The example.init is null",
            )
            assertEquals(
                initJsonArray.size,
                parameterTypeNames.size,
                "The example.init.size != parameter.size $initJsonArray $parameterTypeNames",
            )
            return ExampleSet.json2array(initJsonArray, parameterTypeNames)
        }
    }
}
